test/python/circuit/test_controlled_gate.py

```diff
@@ -0,0 +1,402 @@
+
+    @data(1, 2, 3, 4)
+    def test_inverse_x(self, num_ctrl_qubits):
+        """test inverting ControlledGate"""
+        cnx = XGate().q_if(num_ctrl_qubits)
+        inv_cnx = cnx.inverse()
+        result = Operator(cnx).compose(Operator(inv_cnx))
+        np.testing.assert_array_almost_equal(result.data,
+                                             np.identity(result.dim[0]))
+
+    @data(1, 2, 3)
+    def test_inverse_circuit(self, num_ctrl_qubits):
+        """test inverting ControlledGate"""
+        qc = QuantumCircuit(3)
+        qc.h(0)
+        qc.cx(0, 1)
+        qc.cx(1, 2)
+        qc.rx(np.pi/4, [0, 1, 2])
+        gate = instruction_to_gate(qc.to_instruction())
+        cgate = gate.q_if(num_ctrl_qubits)
+        inv_cgate = cgate.inverse()
+        result = Operator(cgate).compose(Operator(inv_cgate))
+        np.testing.assert_array_almost_equal(result.data,
+                                             np.identity(result.dim[0]))
+
+    @data(1, 2, 3, 4, 5)
+    def test_controlled_unitary(self, num_ctrl_qubits):
+        """test controlled unitary"""
+        num_target = 1
+        n_qubits = num_ctrl_qubits + num_target
+        q_control = QuantumRegister(num_ctrl_qubits)
+        q_target = QuantumRegister(num_target)
+        qc1 = QuantumCircuit(q_target)
+        # for h-rx(pi/2)
+        theta, phi, lamb = 1.57079632679490, 0.0, 4.71238898038469
+        qc1.u3(theta, phi, lamb, q_target[0])
+        base_gate = instruction_to_gate(qc1.to_instruction())
+        # get UnitaryGate version of circuit
+        base_op = Operator(qc1)
+        base_mat = base_op.data
+        cgate = base_gate.q_if(num_ctrl_qubits)
+        test_op = Operator(cgate)
+        cop_mat = _compute_control_matrix(base_mat, num_ctrl_qubits)
+        self.assertTrue(is_unitary_matrix(base_mat))
+        self.assertTrue(matrix_equal(cop_mat, test_op.data, ignore_phase=True))
+
+    @data(1, 2, 3)
+    def test_global_phase(self, num_ctrl_qubits):
+        """test controlled unitary"""
+        # create unitary gate
+        #mat1 = np.array([[1, 0], [0, -1]], dtype=float)
+        mat1 = np.array([[0, 1], [1, 0]], dtype=float)
+        mat2 = np.exp(1j) * mat1
+        gate1 = UnitaryGate(mat1)
+        gate2 = UnitaryGate(mat2)
+        cgate1 = gate1.q_if(num_ctrl_qubits)
+        cgate2 = gate2.q_if(num_ctrl_qubits)
+        cop_mat1 = _compute_control_matrix(mat1, num_ctrl_qubits)
+        cop_mat2 = _compute_control_matrix(mat2, num_ctrl_qubits)
+        n_qubits = 1 + num_ctrl_qubits
+        self.assertTrue(is_unitary_matrix(mat1))
+        self.assertTrue(is_unitary_matrix(mat2))
+        self.assertTrue(is_unitary_matrix(cop_mat1))
+        self.assertTrue(is_unitary_matrix(cop_mat2))
+        self.assertTrue(Operator(cgate1).equiv(Operator(cgate2)))
+        # not sure why following fails
+        #self.assertTrue(matrix_equal(cop_mat1, cop_mat2, ignore_phase=True))
+
+
+def _compute_control_matrix(base_mat, num_ctrl_qubits):
```

Comment:
would it be useful to write this as the `to_matrix()` method of the ControlledGate class?

Reply:
Is that already available with `Operator(Controlledgate).data`?